### DISCUSSION TITLE
"Prop `href` did not match.", when using next/link from external package.

### DISCUSSION THREAD

**Discussion by djfarly** [vercel/next.js#33605] [Ideas]
### Run `next info` (available from version 12.0.8 and up)

I don't think that works on codesandbox: `> No such directory exists as the project root: /sandbox/info`, sorry 😢 

### What version of Next.js are you using?

12.0.9

### What version of Node.js are you using?

14.18.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

any development mode / codesandbox

### Describe the Bug

In short:
I get a rehydration warning like this, when using next/link through any external package and use locales:
```
Warning: Prop `href` did not match. Server: "/my-page" Client: "/fr/my-page"
```

Where I think the problem is:
When a i18n enabled next app uses `next/link` through an external package (that somehow enriches its behaviour), the environment variable `process.env.__NEXT_I18N_SUPPORT` is not defined (only on the server) within the `addLocale` function used by the link.

Thus the link has no locale added on the server. On the client the locale attribute is present as expected. When react rehydrates the html it stumbles upon the a tag with mismatched href attribute.

When not using dev mode the warning is suppressed but I have a hunch that the mismatch is still there.

Also see the discussion here: #22130 (esp. some of my strange logging adventures [here](https://github.com/vercel/next.js/issues/22130#issuecomment-833610774))

### Expected Behavior

`process.env.__NEXT_I18N_SUPPORT` should be defined within `next/link` no matter if server or client and no matter if imported directly or via an external package. Otherwise it would be impossible to write packages abstracting link related things.

In other words using `next/link` through external packages should not cause a warning and should have correct html generated on the server.

### To Reproduce

I have created a reproduction on **[Codesandbox](https://codesandbox.io/s/nextjs-prop-href-did-not-match-9ob4o?file=/pages/index.js)**. Just check `pages/index.js` and the console.

I published a package that does almost nothing but using `next/link` to create this reproduction (it only has `next` and `react` as peerDeps). Check the source here: https://unpkg.com/next-link-30d@1.0.0/index.js

<img width="730" alt="Bildschirmfoto 2022-01-23 um 23 17 43" src="https://user-images.githubusercontent.com/5230863/150700171-c037aa48-aeb5-47da-b83f-59e1fac72476.png">

I would love if you could help me out. This bug has been haunting me for half a year now and I finally managed to reproduce it reliably. ❤️

**Comment by balazsorban44** [comment]
Hi, thank you for raising this. This is currently not meant to work, so I moved this to a discussion/Idea.

As a workaround, I _think_ your distributed component could take `next/link`'s `Link` component as a prop for now. :thinking:

**Reply by djfarly** [reply]
Hej, thx for your reply!

To me it feels kind of unintuitive that it does not work. Thus leading to hard to track down bugs. (Even though I absolutely understand that it's an uncommon use case.)
Is there a more detailed explanation to why it is not supported or is it just "how-webpack-works-unfortunately"?

My use case is a library that connects to a [data-source](https://www.sanity.io/) in getStaticProps and renders "page builder"-style content automatically.

It looks something like this (`pages/index.js`):

```
export { getStaticProps } from '@lcifr/pages/getStatic';
import { PageContent } from '@lcifr/components/page-content';
import { Header } from 'components/Header';

export default function Index() {
  return (
    <>
      <Header />
      <PageContent />
    </>
  );
}
```

Somewhere inside `<PageContent />` could potentially be a link (or not).

It would be a solution to do `<PageContent Link={Link} />` and then store the component in context behind the scenes to use it somewhere in the tree.

But tbh this feels like a very strange API for the user to implement. It's far from intuitive why this next component is passed here.

Another question that comes to my mind: Is the same true for other next components like `next/image`? Does it also use `env` variables that I just didn't notice creating problems behind the scenes?

tyvm 🙏

**Reply by intensr** [reply]
@balazsorban44 could you share some insights based on djfarly's questions?

**Reply by ochicf** [reply]
Hi, I'm experiencing this issue and I have attempted the proposed solution here without any success.

Some context:
- We have a package of components that use a `Link` component that it's a wrapper of a `LinkComponent` that has the same interface of next/link. This `LinkComponent` is read by our `Link` from context, and by default not import next/link (it's supposed to work outside next applications), it's just a simplified version of it that we own.
- The package exposes a `NextProvider` that will import next/link and set it as the value of the `LinkComponent` provider. We use this provider to wrap the react tree in the `_app.tsx` file.
- Before i18n everything worked well, but now we experience this missmatch in paths between server and client.
- We use mui/v5 so we render the link as `<MuiLink component={LinkComponent} />`, but I have tried removing this bit without any change in behaviour. I've also followed the props in React devtools and they arrive as expected to next/link (both `href` and `locale` are the expected values)

I have tried the solution by passing the next/link to our `NextProvider` so it's imported from the project instead of the package (as mentioned in this thread as potential solution), but the behaviour is the same.

<details>
<summary>
To expand and illustrate the above, let me show some code (simplified versions of it):
</summary>

```tsx
// @my/package/Link.tsx

export default function Link(props) {
  const LinkComponent = useContext(LinkComponentContext);
  return <MuiLink {...props} component={LinkComponent} />;
  // same happens with
  return <LinkComponent {...props} />;
}
```
```tsx
// @my/package/next/NextProvider.tsx`
import NextLink from 'next/link.js';

export default function NextProvider({ Link = NextLink, children }) {
  return (
    <LinkComponent.Provider value={Link}>
      {children}
    </LinkComponent.Provider>
  );
}
```
```tsx
// project/pages/_app.tsx
import NextLink from 'next/link.js';

export default function MyApp({ Component, pageProps }) {
  return (
    // experiencing issue with both of following ways
    // <NextProvider Link={NextLink} />
    <NextProvider>
       <Component {...pageProps} />
    </NextProvider>
  );
}

```
</details>

**Comment by ochicf** [comment]
After [sending this comment](https://github.com/vercel/next.js/discussions/33605#discussioncomment-4993739), I tried adding `__NEXT_I18N_SUPPORT=true` in my `.env` file. It seems that this makes the next/link imported from a package to properly read the value for this variable. Looked at it briefly but seems to fix the issue both in dev and production modes.

I guess this variable is internal from Next, so is setting this variable something we should not do? Any side effects or unexpected behaviour I could experience by doing this?
